### PR TITLE
Graph duplicate checks

### PR DESCRIPTION
**Pull request by dcoutts**
This is a continuation and update of #4027 (which got closed automatically for some reason).

Change Graph.fromList to fromDistinctList and fix consequences.

It's really an error to try and build a graph where you have duplicate node keys, so remove Graph.fromList and add Graph.fromDistinctList. This check is always on, not just an assertion, becuase we get it for free given the way Maps can be constructed.

All uses of Graph.fromList are now ok to convert to fromDistinctList, since #4038 and #4039 were merged.

### CODE REVIEW

File: Cabal/Distribution/Compat/Graph.hs [hunk 1]
```diff
@@ -368,12 +372,14 @@ fromMap m
...
     bounds = (0, Map.size m - 1)
 
--- | /O(V log V)/. Convert a list of nodes into a graph.
-fromList :: IsNode a => [a] -> Graph a
-fromList ns = fromMap
-            . Map.fromList
-            . map (\n -> n `seq` (nodeKey n, n))
-            $ ns
+-- | /O(V log V)/. Convert a list of nodes (with distinct keys) into a graph.
+fromDistinctList :: (IsNode a, Show (Key a)) => [a] -> Graph a
```
Comment (ezyang) [inline Cabal/Distribution/Compat/Graph.hs]: I still want this type to use WithCallStack!

File: cabal-install/Distribution/Solver/Modular.hs [hunk 2]
```diff
@@ -53,8 +52,7 @@ modularResolver sc (Platform arch os) cinfo iidx sidx pkgConfigDB pprefs pcs pns
       -- Results have to be converted into an install plan. 'convCP' removes
       -- package qualifiers, which means that linked packages become duplicates
       -- and can be removed.
-      -- TODO: Use ordNubBy instead of nubBy.
-      postprocess a rdm = nubBy ((==) `on` nodeKey) $
+      postprocess a rdm = ordNubBy nodeKey $
```
Comment (grayjay) [inline cabal-install/Distribution/Solver/Modular.hs]: Thanks!

**Comment by dcoutts** [comment]
@ezyang I'm sympathetic to wanting to use WithCallStack, but currently doing so does not provide the results we would want:
```
Graph.fromDistinctList: duplicate key: UnitId
"Cabal-1.24.1.0-041ce14db104a940f73b1dff57ff4ac49c05649dc35062bb52db33b613e06f52"
?callStack, called at ./Distribution/Compat/Stack.hs:42:13 in CgUwfkcqK4d8RNzmWp2WZa:Distribution.Compat.Stack
callStack, called at ./Distribution/Compat/Graph.hs:384:48 in CgUwfkcqK4d8RNzmWp2WZa:Distribution.Compat.Graph
```
It doesn't give use the caller of `Graph.fromDistinctList`, which is the only thing we're actually interested in.

So I suggest we just merge as is, and solve the callstack problem separately.

**Comment by ezyang** [comment]
Yes, OK, I see the behavior you're describing. Let's go and merge then (but after fixing the conflict!)

**Comment by 23Skidoo** [comment]
@dcoutts Can you please fix the merge conflict so that we could merge this?

**Comment by ezyang** [comment]
Subsumed by #4300

**Pull request closed without merging**